src/elastio-snap.c

```diff
@@ -1775,9 +1798,225 @@ static int file_io(struct file *filp, int is_write, void *buf, sector_t offset,
 
 	return 0;
 }
+
 #define file_write(filp, buf, offset, len) file_io(filp, 1, buf, offset, len)
 #define file_read(filp, buf, offset, len) file_io(filp, 0, buf, offset, len)
 
+#define SECTOR_INVALID ~(u64)0
+
+sector_t sector_by_offset(struct snap_device *dev, size_t offset)
+{
+	unsigned int i;
+	struct fiemap_extent *extent = dev->sd_cow_extents;
+	for (i = 0; i < dev->sd_cow_ext_cnt; i++) {
+		// TODO: double check if offset should be strictly less that fe_logical or not
+		if (offset >= extent[i].fe_logical && offset < extent[i].fe_logical + extent[i].fe_length)
+			return (extent[i].fe_physical + offset) >> 9;
+	}
+
+	return SECTOR_INVALID;
+}
+
+int file_write_block(struct snap_device *dev, void *block, size_t offset, size_t len)
+{
+	int ret;
+	int bytes;
+	char *data;
+	struct page *pg;
+	struct bio_set *bs;
+	struct bio *new_bio;
+	struct block_device *bdev;
+	sector_t start_sect;
+	int sectors_processed;
+
+	ret = 0;
+	bs = dev_bioset(dev);
+	bdev = dev->sd_base_dev;
+	sectors_processed = 0;
+
+	WARN_ON(!IS_ALIGNED(offset, PAGE_SIZE) || len > SECTORS_PER_BLOCK);
+
+	if (dev->sd_cow)
+		file_unlock(dev->sd_cow->filp);
+
+write_bio:
+	start_sect = sector_by_offset(dev, offset);
+
+//	new_bio = bio_alloc_bioset(GFP_NOIO, 1, bs);
+	new_bio = bio_alloc(GFP_NOIO, 1);
+	if(!new_bio){
+		ret = -ENOMEM;
+		LOG_ERROR(ret, "error allocating bio (write) - bs = %p", bs);
+		goto out;
+	}
+
+	bio_set_dev(new_bio, bdev);
+	elastio_snap_set_bio_ops(new_bio, REQ_OP_WRITE, 0);
+	bio_sector(new_bio) = start_sect;
+	bio_idx(new_bio) = 0;
+
+	pg = alloc_page(GFP_NOIO);
+	if(!pg){
+		ret = -ENOMEM;
+		LOG_ERROR(ret, "error allocating read bio page");
+		goto out;
+	}
+
+	data = kmap(pg);
+
+	do {
+		int write_offset = sectors_processed * SECTOR_SIZE;
+		memcpy(data + write_offset, block + write_offset, SECTOR_SIZE);
+		offset += SECTOR_SIZE;
+		sectors_processed++;
+	} while (sectors_processed < len &&
+			sector_by_offset(dev, offset) == start_sect + sectors_processed);
+
+	kunmap(pg);
+
+	bytes = bio_add_page(new_bio, pg, PAGE_SIZE, 0);
+	if(bytes != PAGE_SIZE){
+		LOG_DEBUG("bio_add_page() error!");
+		__free_page(pg);
+		ret = -EFAULT;
+		goto out;
+	}
+
+	pg->mapping = dev->sd_cow_inode->i_mapping;
+
+	ret = submit_bio_wait(new_bio);
```

Comment:
Because we are waiting until the blocks are read/written. For 'write' this is important because if we exit instantaneously, some further logic, which relies on the written data, can be broken.